backend/python-functions/data_ingest.py

```diff
@@ -0,0 +1,145 @@
+    2020-06-04,State Prisons,Colorado,Denver Reception & Diagnostic Center,253,2,,0,,,,,,,
+
+    Would be reshaped to:
+
+    {
+      "Colorado::Denver Reception & Diagnostic Center": {
+        "canonicalName": "Denver Reception & Diagnostic Center",
+        "facilityType": "State Prisons",
+        "state": "Colorado",
+        "covidCases": {
+          "2020-06-01": {
+            "popDeaths": 0,
+            "popTested": 253,
+            "popTestedPositive": 2
+          },
+          "2020-06-02": {
+            "popDeaths": 0,
+            "popTested": 253,
+            "popTestedPositive": 2
+          },
+          "2020-06-03": {
+            "popDeaths": 0,
+            "popTested": 253,
+            "popTestedPositive": 2
+          },
+          "2020-06-04": {
+            "popDeaths": 0,
+            "popTested": 253,
+            "popTestedPositive": 2
+          }
+        }
+      },
+      ...
+    }
+    """
+    facilities = {}    
+    with open(file_location) as csv_file:
+        reader = csv.DictReader(csv_file, delimiter=',')
+
+        for row in reader:
+            state = row["State"].strip()
+            canonical_facility_name = row["Canonical Facility Name"].strip()
+            facility_type = row['Facility Type'].strip()
+            key = f'{state}::{canonical_facility_name}'
+            
+            if(key in facilities):
+                # If the facility already exists, append the case counts for a given day.
+                facilities[key]['covidCases'][row['Date']] = build_covid_case_counts(row)
+            else:
+                # If the facility does not already exist, save its metadata along with
+                # the case counts for a given day.
+                facilities[key] = {
+                    'canonicalName': canonical_facility_name,
+                    'facilityType': facility_type,
+                    'state': state,
+                    'covidCases': {
+                        f'{row["Date"]}': build_covid_case_counts(row)
+                    }
+                }
+
+    return facilities
+
+def persist(facilities):
+    db = firestore.Client()
+
+    facilities_collection = db.collection(REFERENCE_FACILITIES_COLLECTION_ID)
+                
+    for key, facility in facilities.items():
+        facilitiesQueryResult = facilities_collection \
+            .where('state', '==', f'{facility["state"]}') \
+            .where('canonicalName', '==', f'{facility["canonicalName"]}') \
+            .stream()
+
+        facilityDocuments = list(facilitiesQueryResult)       
+
+        if len(facilityDocuments) == 0 or len(facilityDocuments) == 1:
+            # Accesses an existing reference facility document or creates a new one if
+            # one does not exist.
+            facility_id = facilityDocuments[0].id if len(facilityDocuments) == 1 else None
+            facility_ref = facilities_collection.document(facility_id)
```

Comment:
What happens here if `facility_id` is None?

Reply:
If `facility_id` is none then we'll create a new facility document.  Essentially lines 122 and 123 determine whether or not we're performing a create or an update.